tests/tt_eager/python_api_testing/unit_testing/misc/test_nlp_kv_cache_load_slice.py

```diff
@@ -0,0 +1,135 @@
+# SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+
+# SPDX-License-Identifier: Apache-2.0
+
+import pytest
+from loguru import logger
+
+import numpy as np
+import torch
+
+import tt_lib as ttl
+from models.utility_functions import is_grayskull, comp_pcc
+
+
+def unpadding_test(
+    kv_cache_shape,
+    seq_len_start,
+    seq_len_end,
+    device,
+    dtype,
+):
+    if dtype == ttl.tensor.DataType.FLOAT32:
+        inp = torch.rand(*kv_cache_shape, dtype=torch.float)
+    else:
+        inp = torch.rand(*kv_cache_shape, dtype=torch.bfloat16)
+
+    test_tensor = (
+        ttl.tensor.Tensor(
+            inp.reshape(-1).tolist(),
+            inp.shape,
+            dtype,
+            ttl.tensor.Layout.ROW_MAJOR,
+        )
+        .to(ttl.tensor.Layout.TILE)
+        .to(device)
+    )
+    # breakpoint()
+    test_tensor_tt = ttl.tensor.nlp_kv_cache_load_slice(test_tensor, seq_len_start, seq_len_end)
+
+    test_tensor_pt = test_tensor_tt.cpu().to(ttl.tensor.Layout.ROW_MAJOR).to_torch()
+
+    # Pytorch reference
+    test_tensor_ref = inp[:, :, seq_len_start:seq_len_end]
+
+    return test_tensor_pt, test_tensor_ref, test_tensor_tt.memory_config()
+
+
+@pytest.mark.parametrize(
+    "dtype",
+    (ttl.tensor.DataType.BFLOAT8_B, ttl.tensor.DataType.BFLOAT16, ttl.tensor.DataType.FLOAT32),
+    ids=["bfloat8_b", "bfloat16", "float"],
+)
+@pytest.mark.parametrize(
+    "kv_cache_shape, seq_len_start, seq_len_end",
+    (
+        ((9, 1, 64, 64), 0, 32),
+        ((9, 2, 64, 64), 0, 32),
+        ((9, 3, 64, 64), 32, 64),
+        ((3, 2, 64, 64), 32, 64),
+        ((1, 1, 64, 64), 32, 64),
+        ((1, 1, 128, 96), 32, 64),
+        ((1, 1, 128, 96), 64, 96),
+        ((1, 3, 32, 32), 0, 32),
+        ((1, 6, 32, 32), 0, 32),
+        ((1, 6, 128, 64), 32, 128),
+        ((4, 6, 128, 64), 96, 128),
+        ((1, 32, 2048, 64), 0, 1024),
+        ((1, 32, 2048, 64), 128, 1024),
+        ((32, 1, 2048, 64), 1024, 1056),
+        ((32, 1, 2048, 64), 0, 2016),
+        ((32, 1, 2048, 64), 0, 2048),
+        ((32, 1, 2048 + 128, 128), 0, 2048),  # llama2 70B use case
+        ((1, 32, 2048 + 128, 128), 0, 2048),  # llama2 70B use case
+    ),
+)
+def test_run_unpadding_test(
+    kv_cache_shape,
+    seq_len_start,
+    seq_len_end,
+    device,
+    dtype,
+    use_program_cache,
+):
+    if is_grayskull():
+        pytest.skip("Skipping test on Grayskull")
+
+    for i in range(3):
+        a_pt, a_ref, memory_config = unpadding_test(
+            kv_cache_shape,
+            seq_len_start,
+            seq_len_end,
+            device,
+            dtype,
+        )
```

Comment:
what does this `for` loop do? If you are testing program cache and callback, you should:
- Have the input/output buffers be shifted so you actually test the runtime arg update of the callback
- Assert on number of expected cache entries

Reply:
Depends on your test and how deallocation happens. But you can achieve this by:
- creating a very small (eg. single tile) tensor in between your loops. This way when the test runs, the base addresses of your input/output tensors are different

Reply:
Gotcha!